lib/benchmark/linear.go

```diff
@@ -0,0 +1,79 @@
+package benchmark
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/trace"
+)
+
+// Linear generator
+type Linear struct {
+	LowerBound          int
+	UpperBound          int
+	Step                int
+	MinimumMeasurements int
+	MinimumWindow       time.Duration
+	Threads             int
+	currentRPS          int
+	config              Config
+}
+
+// Generate advances the Generator to the next generation.
+func (lg *Linear) Generate() bool {
+	if lg.currentRPS < lg.LowerBound {
+		lg.currentRPS = lg.LowerBound
+		return true
+	}
+	lg.currentRPS += lg.Step
+	return lg.currentRPS <= lg.UpperBound
+}
+
+// GetBenchmark returns the benchmark config for the current generation.
+func (lg *Linear) GetBenchmark() (context.Context, Config, error) {
+	if lg.currentRPS > lg.UpperBound {
+		return nil, Config{}, errors.New("No more generations")
+	}
+
+	return context.TODO(), Config{
+		MinimumWindow:       lg.MinimumWindow,
+		MinimumMeasurements: lg.MinimumMeasurements,
+		Rate:                lg.currentRPS,
+		Threads:             lg.Threads,
+		Command:             lg.config.Command,
+	}, nil
+}
+
+// Benchmark runs the benchmark of receiver type
+func (lg *Linear) Benchmark(ctx context.Context, command []string, tc *client.TeleportClient) ([]Result, error) {
+	var result Result
+	var results []Result
+	sleep := false
+
+	for lg.Generate() {
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
```

Comment:
`defer` statements run at the end of the function, not the end of the loop.  If you want to cancel the context on each iteration, you'll need to do it manually.  If you wan't to simply cancel on function exit (current behavior), you can achieve this by deferring a single cancel before the start of the loop.